src/utils/web_byte_utils.ts

```diff
@@ -0,0 +1,138 @@
+  new (): TextEncoder;
+};
+
+// Web global
+declare const TextDecoder: TextDecoderConstructor;
+declare const TextEncoder: TextEncoderConstructor;
+declare const atob: (base64: string) => string;
+declare const btoa: (binary: string) => string;
+
+type ArrayBufferViewWithTag = ArrayBufferView & {
+  [Symbol.toStringTag]?: string;
+};
+
+const HEX_DIGIT = /(\d|[a-f])/i;
+
+export const webByteUtils = {
+  toLocalBufferType(
+    potentialUint8array: Uint8Array | ArrayBufferViewWithTag | ArrayBuffer
+  ): Uint8Array {
+    let stringTag = potentialUint8array?.[Symbol.toStringTag];
+    if (typeof stringTag !== 'string') {
+      stringTag = Object.prototype.toString.call(potentialUint8array);
+      stringTag = stringTag.slice(8, stringTag.indexOf(']'));
+    }
+
+    if (stringTag === 'Uint8Array') {
+      return potentialUint8array as Uint8Array;
+    }
+
+    if (ArrayBuffer.isView(potentialUint8array)) {
+      return new Uint8Array(
+        potentialUint8array.buffer.slice(
+          potentialUint8array.byteOffset,
+          potentialUint8array.byteOffset + potentialUint8array.byteLength
+        )
+      );
+    }
+
+    if (stringTag === 'ArrayBuffer' || stringTag === 'SharedArrayBuffer') {
+      return new Uint8Array(potentialUint8array);
+    }
+
+    throw new BSONError(`Cannot make a Uint8Array from ${String(potentialUint8array)}`);
+  },
+
+  allocate(size: number): Uint8Array {
+    if (typeof size !== 'number') {
+      throw new TypeError(`The "size" argument must be of type number. Received ${String(size)}`);
+    }
+    return new Uint8Array(size);
+  },
+
+  equals(a: Uint8Array, b: Uint8Array): boolean {
+    if (a.byteLength !== b.byteLength) {
+      return false;
+    }
+    for (let i = 0; i < a.byteLength; i++) {
+      if (a[i] !== b[i]) {
+        return false;
+      }
+    }
+    return true;
+  },
+
+  fromNumberArray(array: number[]): Uint8Array {
+    return Uint8Array.from(array);
+  },
+
+  fromBase64(base64: string): Uint8Array {
+    return Uint8Array.from(atob(base64), c => c.charCodeAt(0));
+  },
+
+  toBase64(uint8array: Uint8Array): string {
+    return btoa(webByteUtils.toISO88591(uint8array));
+  },
+
+  fromISO88591(codePoints: string): Uint8Array {
+    return Uint8Array.from(codePoints, c => c.charCodeAt(0) & 0xff);
+  },
+
+  toHex(uint8array: Uint8Array): string {
+    return Array.from(uint8array, byte => byte.toString(16).padStart(2, '0')).join('');
+  },
+
+  fromUTF8(text: string): Uint8Array {
+    return new TextEncoder().encode(text);
+  },
+
+  toISO88591(uint8array: Uint8Array): string {
+    return Array.from(Uint16Array.from(uint8array), b => String.fromCharCode(b)).join('');
+  },
+
+  fromHex(hex: string): Uint8Array {
+    const evenLengthHex = hex.length % 2 === 0 ? hex : hex.slice(0, hex.length - 1);
+    const buffer = [];
+
+    for (let i = 0; i < evenLengthHex.length; i += 2) {
+      if (!HEX_DIGIT.test(evenLengthHex[i]) && !HEX_DIGIT.test(evenLengthHex[i + 1])) {
```

Comment:
is this the same as the node internal implementation? to silently ignore any non-hex portions or whole non-hex strings?

Reply:
Yep I should have tests to demonstrate this, I actually didn't look at the node implementation, but the tests demonstrate it matches, node always ignores the last character in an uneven lengthed input. And Node will stop short when it reaches a character that is not hex. I assume the idea is to mirror parseInt, which does similiar early stoping.